internal/cluster/store/fsm.go

```diff
@@ -0,0 +1,134 @@
+package store
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"github.com/hashicorp/raft"
+	"github.com/pbinitiative/zenbpm/internal/cluster/command/proto"
+	pb "google.golang.org/protobuf/proto"
+)
+
+// FSM is Finite State Machine of the system state
+type FSM struct {
+	s *Store
+}
+
+// NewFSM returns a new FSM.
+func NewFSM(s *Store) *FSM {
+	return &FSM{s: s}
+}
+
+var _ raft.FSM = &FSM{}
+
+// Apply applies a Raft log entry to the key-value store.
+func (f *FSM) Apply(l *raft.Log) interface{} {
+	f.s.mu.Lock()
+	defer f.s.mu.Unlock()
+	var command proto.Command
+	if err := pb.Unmarshal(l.Data, &command); err != nil {
+		panic(fmt.Sprintf("failed to unmarshal command: %s", err.Error()))
+	}
+
+	switch command.GetType() {
+	case proto.Command_TYPE_NODE_CHANGE:
+		nodeChangeCommand := command.GetNodeChange()
+		return f.applyNodeChange(nodeChangeCommand)
+	case proto.Command_TYPE_NODE_PARTITION_CHANGE:
+		partitionChangeCommand := command.GetNodePartitionChange()
+		return f.applyPartitionChange(partitionChangeCommand)
+	default:
+		panic(fmt.Sprintf("unrecognized command type: %s", command.Type))
+	}
+}
+
+func (f *FSM) Snapshot() (raft.FSMSnapshot, error) {
+	f.s.mu.Lock()
+	defer f.s.mu.Unlock()
+
+	return &fsmSnapshot{ClusterState: *f.s.state.DeepCopy()}, nil
+}
+
+func (f *FSM) Restore(rc io.ReadCloser) error {
+	snapshot := fsmSnapshot{}
+	if err := json.NewDecoder(rc).Decode(&snapshot); err != nil {
+		return err
+	}
+
+	// Set the state from the snapshot, no lock required according to
+	// Hashicorp docs.
```

Comment:
Could you please share an URL/reference to the docs - that seems to be very useful?

Reply:
https://github.com/hashicorp/raft/blob/main/fsm.go#L43